src/app/ConcreteAttributePath.h

```diff
@@ -138,6 +145,14 @@ struct ConcreteDataAttributePath : public ConcreteAttributePath
                         ChipLogValueMEI(mClusterId), ChipLogValueMEI(mAttributeId));
     }
 
+    bool operator==(const ConcreteDataAttributePath & aOther) const
```

Comment:
Just take a further look at the test failure in Linux. it fails here, https://github.com/project-chip/connectedhomeip/blob/master/src/app/ReadClient.cpp#L803
. Previously operator in base class is inherited by derived one. It is just comparing the attribute path without data version thing, when receiving ICD operating mode attribute from report, that data attribute path has version inside, we expect to compare directly against the attribute path operating mode attribute without  version. With your proposed change, "==" would fail because of data version. You may update readClient with the below.

--- a/src/app/ReadClient.cpp
+++ b/src/app/ReadClient.cpp
@@ -799,8 +799,8 @@ CHIP_ERROR ReadClient::ProcessAttributeReportIBs(TLV::TLVReader & aAttributeRepo
                 attributePath.mListOp = ConcreteDataAttributePath::ListOperation::ReplaceAll;
             }
 
-            if (attributePath ==
-                ConcreteDataAttributePath(kRootEndpointId, Clusters::IcdManagement::Id,
+            ConcreteAttributePath & path = attributePath;
+            if (path == ConcreteAttributePath(kRootEndpointId, Clusters::IcdManagement::Id,
                                           Clusters::IcdManagement::Attributes::OperatingMode::Id))
             {
                 PeerType peerType;

However, if going with this PR, the concerning thing is to audit ConcreteReadAttributePath and ConcreteDataAttributePath usage across the whole sdk, underlying == comparator in based one may have been broadly used without enough testing(CI cannot capture) and even manufacturer's server code may use them..

Reply:
Adding this operator== requires a _really_ careful audit of every single use of ConcreteDataAttributePath.  Was this audit done in the process of creating this PR @mwswartwout (and this spot was missed)?  Or was there no audit at all?

Reply:
I didn't manually audit the SDK to find all uses of the equality operator.

Does that seem like a sane thing to do? Or should I just introduce a StrictEquals function? This does seem like a huge gotcha that could cause bugs in the future, but I'm not sure how widely existing code is depending on this behavior.

Reply:
You'd need to check uses of ConcreteDataAttributePath, not all uses of the equality operator.

I agree that there are various gotchas here.  Ideally operator== would in fact test for equality, and any code depending on the current behavior where two ConcreteDataAttributePath are considered equal if they have the same attribute path would be fixed.

So I think the sanest option here is:

1. Add a `bool ConcreteDataAttributePath::MatchesAttributePath(const ConcreteAttributePath &)` which just tests the endpoint/cluster/attribute (possibly by calling the superclass operator==?).
2. Do the audit, convert things to use that instead of operator== as needed.
3. Change the behavior of operator==.

Reply:
We may prohibit operator ==  and check how many failure we can see in all builds..

Reply:
I deleted the operator== from both ConcreteReadAttributePath and ConcreteDataAttributePath and the only build failure that occurred was in ReadClient.cpp.

Reply:
Awesome, thanks

Reply:
I created https://github.com/project-chip/connectedhomeip/pull/32131with only the operator deletion. We should merge that first (since I only tested the Linux build locally and need to rely on CI to catch other usages of the equality operator), and then we can merge this PR as just adding the new equality operators.